### ISSUE TITLE
io_uring test failure

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#10247]
### Zig Version

0.9.0-dev.1812+1ea650bb7

### Steps to Reproduce

```
$ uname -a 
Linux ark 5.4.122 #1-NixOS SMP Wed May 26 10:05:21 UTC 2021 x86_64 GNU/Linux
$ ./zig test ../lib/std/std.zig
```

### Expected Behavior

All std lib tests passed.

### Actual Behavior

After #10151 was merged, this started happening:

```
Test [293/2086] os.linux.io_uring.test "timeout_link_chain1"... Req 0x22222222 got -9
Test [294/2086] os.linux.io_uring.test "timeout_link_chain1"... FAIL (TestUnexpectedResult)
/home/andy/Downloads/zig/lib/std/io/writer.zig:23:26: 0x5d84b1 in io.writer.Writer(*io.fixed_buffer_stream.FixedBufferStream([]u8),io.fixed_buffer_stream.WriteError,io.fixed_buffer_stream.FixedBufferStream([]u8).write).writeAll (test)
                index += try self.write(bytes[index..]);
                         ^
/home/andy/Downloads/zig/lib/std/io/fixed_buffer_stream.zig:63:46: 0x5d801a in io.fixed_buffer_stream.FixedBufferStream([]u8).write (test)
            if (self.pos >= self.buffer.len) return error.NoSpaceLeft;
                                             ^
/home/andy/Downloads/zig/lib/std/io/writer.zig:17:13: 0x5d85a2 in io.writer.Writer(*io.fixed_buffer_stream.FixedBufferStream([]u8),io.fixed_buffer_stream.WriteError,io.fixed_buffer_stream.FixedBufferStream([]u8).write).write (test)
            return writeFn(self.context, bytes);
            ^
/home/andy/Downloads/zig/lib/std/io/writer.zig:23:26: 0x5d84b1 in io.writer.Writer(*io.fixed_buffer_stream.FixedBufferStream([]u8),io.fixed_buffer_stream.WriteError,io.fixed_buffer_stream.FixedBufferStream([]u8).write).writeAll (test)
                index += try self.write(bytes[index..]);
                         ^
/home/andy/Downloads/zig/lib/std/io/fixed_buffer_stream.zig:63:46: 0x5d801a in io.fixed_buffer_stream.FixedBufferStream([]u8).write (test)
            if (self.pos >= self.buffer.len) return error.NoSpaceLeft;
                                             ^
/home/andy/Downloads/zig/lib/std/io/writer.zig:17:13: 0x5d85a2 in io.writer.Writer(*io.fixed_buffer_stream.FixedBufferStream([]u8),io.fixed_buffer_stream.WriteError,io.fixed_buffer_stream.FixedBufferStream([]u8).write).write (test)
            return writeFn(self.context, bytes);
            ^
/home/andy/Downloads/zig/lib/std/io/writer.zig:23:26: 0x5d84b1 in io.writer.Writer(*io.fixed_buffer_stream.FixedBufferStream([]u8),io.fixed_buffer_stream.WriteError,io.fixed_buffer_stream.FixedBufferStream([]u8).write).writeAll (test)
                index += try self.write(bytes[index..]);
                         ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:39:27: 0x77ffaf in os.linux.io_uring.IO_Uring.init_params (test)
        if (entries == 0) return error.EntriesZero;
                          ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:31:16: 0x663f3d in os.linux.io_uring.IO_Uring.init (test)
        return try IO_Uring.init_params(entries, &params);
               ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:40:46: 0x77fff6 in os.linux.io_uring.IO_Uring.init_params (test)
        if (!std.math.isPowerOfTwo(entries)) return error.EntriesNotPowerOfTwo;
                                             ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:31:16: 0x663f3d in os.linux.io_uring.IO_Uring.init (test)
        return try IO_Uring.init_params(entries, &params);
               ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:143:46: 0x7818e9 in os.linux.io_uring.IO_Uring.get_sqe (test)
        if (next -% head > self.sq.sqes.len) return error.SubmissionQueueFull;
                                             ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:354:21: 0x664228 in os.linux.io_uring.IO_Uring.nop (test)
        const sqe = try self.get_sqe();
                    ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:1504:36: 0x46e7ae in os.linux.io_uring.test "write/read" (test)
    if (cqe_write.err() == .INVAL) return error.SkipZigTest;
                                   ^
/home/andy/Downloads/zig/lib/std/os.zig:2127:19: 0x9cfe9f in os.unlinkatZ (test)
        .NOENT => return error.FileNotFound,
                  ^
/home/andy/Downloads/zig/lib/std/fs.zig:1633:25: 0x782f9b in fs.Dir.deleteFileZ (test)
            else => |e| return e,
                        ^
/home/andy/Downloads/zig/lib/std/fs.zig:1612:13: 0x6654c7 in fs.Dir.deleteFile (test)
            return self.deleteFileZ(&sub_path_c);
            ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:1616:9: 0x470312 in os.linux.io_uring.test "openat" (test)
        return error.SkipZigTest;
        ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:1646:36: 0x470b05 in os.linux.io_uring.test "close" (test)
    if (cqe_close.err() == .INVAL) return error.SkipZigTest;
                                   ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:1686:37: 0x4714b6 in os.linux.io_uring.test "accept/connect/send/recv" (test)
    if (cqe_accept.err() == .INVAL) return error.SkipZigTest;
                                    ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:1827:9: 0x472e6e in os.linux.io_uring.test "timeout_remove" (test)
        return error.SkipZigTest;
        ^
/home/andy/Downloads/zig/lib/std/testing.zig:306:14: 0x5c0d5b in testing.expect (test)
    if (!ok) return error.TestUnexpectedResult;
             ^
/home/andy/Downloads/zig/lib/std/os/linux/io_uring.zig:1892:21: 0x473718 in os.linux.io_uring.test "timeout_link_chain1" (test)
                    try testing.expect(false);
                    ^
/home/andy/Downloads/zig/lib/std/unicode.zig:49:46: 0x8c92af in unicode.utf8Encode (test)
            if (0xd800 <= c and c <= 0xdfff) return error.Utf8CannotEncodeSurrogateHalf;
                                             ^
/home/andy/Downloads/zig/lib/std/unicode.zig:13:5: 0x94eb32 in unicode.utf8CodepointSequenceLength (test)
    return error.CodepointTooLarge;
    ^
/home/andy/Downloads/zig/lib/std/unicode.zig:36:20: 0x8c91f2 in unicode.utf8Encode (test)
    const length = try utf8CodepointSequenceLength(c);
                   ^
/home/andy/Downloads/zig/lib/std/fmt.zig:1769:25: 0x5cf22a in fmt.charToDigit (test)
    if (value >= radix) return error.InvalidCharacter;
                        ^
/home/andy/Downloads/zig/lib/std/fmt.zig:2386:20: 0x5c456d in fmt.hexToBytes (test)
        const lo = try charToDigit(input[in_i + 1], 16);
                   ^
/home/andy/Downloads/zig/lib/std/fmt.zig:2379:9: 0x5c446f in fmt.hexToBytes (test)
        return error.InvalidLength;
        ^
/home/andy/Downloads/zig/lib/std/fmt.zig:2381:9: 0x5c44b3 in fmt.hexToBytes (test)
        return error.NoSpaceLeft;
        ^
/home/andy/Downloads/zig/lib/std/io/fixed_buffer_stream.zig:63:46: 0x5d801a in io.fixed_buffer_stream.FixedBufferStream([]u8).write (test)
            if (self.pos >= self.buffer.len) return error.NoSpaceLeft;
                                             ^
/home/andy/Downloads/zig/lib/std/io/writer.zig:17:13: 0x5d85a2 in io.writer.Writer(*io.fixed_buffer_stream.FixedBufferStream([]u8),io.fixed_buffer_stream.WriteError,io.fixed_buffer_stream.FixedBufferStream([]u8).write).write (test)
            return writeFn(self.context, bytes);
            ^
```

**Comment by hnakamur** [comment]
Sorry but I cannot reproduce the same error.
Actually I got a different error.

I used Zig version 0.9.0-dev.1815+20e19e75f from https://ziglang.org/builds/zig-linux-x86_64-0.9.0-dev.1815+20e19e75f.tar.xz and LXD ubuntu/18.04 virtual machine image.

```
# uname -a
Linux bionic-vm 5.4.0-91-generic #102~18.04.1-Ubuntu SMP Thu Nov 11 14:46:36 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

```
curl -LO https://ziglang.org/builds/zig-linux-x86_64-0.9.0-dev.1815+20e19e75f.tar.xz
tar xf zig-linux-x86_64-0.9.0-dev.1815+20e19e75f.tar.xz
cd zig-linux-x86_64-0.9.0-dev.1815+20e19e75f
```

I got a different error when I ran test for `./lib/std/std.zig`.

```
# ./zig test ./lib/std/std.zig
Semantic Analysis [3438/4623] Killed
# echo $?
137
```

I got no error when I ran test for `./lib/std/os/linux/io_uring.zig`.

```
# ./zig test ./lib/std/os/linux/io_uring.zig
Test [6/17] test "write/read"... SKIP
Test [8/17] test "openat"... SKIP
Test [9/17] test "close"... SKIP
Test [10/17] test "accept/connect/send/recv"... SKIP
Test [13/17] test "timeout_remove"... SKIP
Test [14/17] test "timeout_link_chain1"... SKIP
Test [15/17] test "fallocate"... SKIP
Test [16/17] test "statx"... SKIP
Test [17/17] test "accept/connect/recv/cancel"... SKIP
Test [18/17] test "register_files_update"... SKIP
7 passed; 10 skipped; 0 failed.
# echo $?
0
```

I used a fresh virtual machine created with the following command.

```
lxc init images:ubuntu/18.04/cloud bionic-vm --vm
lxc config device override bionic-vm root size=20GB
lxc start bionic-vm
```

I updated the linux kernel with the following command.

```
lxc exec bionic-vm bash
apt-get install -y linux-image-generic-hwe-18.04
```

**Comment by hnakamur** [comment]
Ah I found you added a workaround at [std: add workaround for failing io_uring test · ziglang/zig@ec10e63](https://github.com/ziglang/zig/commit/ec10e63f4945e81cfc5369a2c886ef0b7ae11c1b).
Thanks.

**Comment by hnakamur** [comment]
The error `Semantic Analysis [3438/4623] Killed` is gone when I increased memory for the virtual machine to 8192MiB with the following command (4096MiB was not enough).

```
lxc stop bionic-vm
lxc config set bionic-vm limits.memory 4096MiB
lxc start bionic-vm
```

I got the following error when I ran `./zig test ./lib/std/std.zig`.

```
./lib/std/compress/gzip.zig:191:20: error: unable to find '/root/zig-linux-x86_64-0.9.0-dev.1815+20e19e75f/lib/std/compress/rfc1952.txt.gz'
        @embedFile("rfc1952.txt.gz"),
                   ^
./lib/std/compress/gzip.zig:191:9: note: referenced here
        @embedFile("rfc1952.txt.gz"),
        ^
./lib/std/compress/zlib.zig:128:20: error: unable to find '/root/zig-linux-x86_64-0.9.0-dev.1815+20e19e75f/lib/std/compress/rfc1951.txt.z.0'
        @embedFile("rfc1951.txt.z.0"),
                   ^
./lib/std/compress/zlib.zig:128:9: note: referenced here
        @embedFile("rfc1951.txt.z.0"),
        ^
./lib/std/crypto/aegis.zig:351:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("79d94593d8c2119d7e8fd9b8fc77845c5c077a05b2528b6ac54b563aed8efe84", &c);
             ^
./lib/std/crypto/aegis.zig:374:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("41de9000a7b5e40e2d68bb64d99ebb19", &c);
             ^
./lib/std/crypto/aegis.zig:391:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("83cc600dc4e3e7e62d4055826174f149", &tag);
             ^
./lib/std/crypto/aegis.zig:407:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("f373079ed84b2709faee373584585d60accd191db310ef5d8b11833df9dec711", &c);
             ^
./lib/std/crypto/aegis.zig:430:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("b98f03a947807713d75a4fff9fc277a6", &c);
             ^
./lib/std/crypto/aegis.zig:447:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("f7a0878f68bd083e8065354071fc27c3", &tag);
             ^
./lib/std/crypto/aes_gcm.zig:113:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("6b6ff610a16fa4cd59f1fb7903154e92", &tag);
             ^
./lib/std/crypto/aes_gcm.zig:125:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("262ed164c2dfb26e080a9d108dd9dd4c", &tag);
             ^
./lib/std/crypto/aes_gcm.zig:141:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("5ca1642d90009fea33d01f78cf6eefaf01d539472f7c", &c);
             ^
./lib/std/crypto/aes_gcm.zig:158:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("5ca1642d90009fea33d01f78cf6eefaf01", &c);
             ^
./lib/std/crypto/blake2.zig:204:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Blake2s160, h1, "");
             ^
./lib/std/crypto/blake2.zig:223:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:275:18: error: container 'crypto.test' has no member called 'assertEqualHash'
        try htest.assertEqualHash(Blake2s160, h1, block[0..]);
                 ^
./lib/std/crypto/blake2.zig:287:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Blake2s224, h1, "");
             ^
./lib/std/crypto/blake2.zig:306:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:357:18: error: container 'crypto.test' has no member called 'assertEqualHash'
        try htest.assertEqualHash(Blake2s224, h1, block[0..]);
                 ^
./lib/std/crypto/blake2.zig:369:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Blake2s256, h1, "");
             ^
./lib/std/crypto/blake2.zig:388:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:425:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:449:18: error: container 'crypto.test' has no member called 'assertEqualHash'
        try htest.assertEqualHash(Blake2s256, h1, block[0..]);
                 ^
./lib/std/crypto/blake2.zig:627:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Blake2b160, h1, "");
             ^
./lib/std/crypto/blake2.zig:646:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:704:18: error: container 'crypto.test' has no member called 'assertEqualHash'
        try htest.assertEqualHash(Blake2b160, h1, block[0..]);
                 ^
./lib/std/crypto/blake2.zig:716:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Blake2b384, h1, "");
             ^
./lib/std/crypto/blake2.zig:735:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:793:18: error: container 'crypto.test' has no member called 'assertEqualHash'
        try htest.assertEqualHash(Blake2b384, h1, block[0..]);
                 ^
./lib/std/crypto/blake2.zig:805:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Blake2b512, h1, "");
             ^
./lib/std/crypto/blake2.zig:824:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:861:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/blake2.zig:885:18: error: container 'crypto.test' has no member called 'assertEqualHash'
        try htest.assertEqualHash(Blake2b512, h1, block[0..]);
                 ^
./lib/std/crypto/md5.zig:240:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Md5, "d41d8cd98f00b204e9800998ecf8427e", "");
             ^
./lib/std/crypto/md5.zig:254:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("d41d8cd98f00b204e9800998ecf8427e", out[0..]);
             ^
./lib/std/crypto/sha1.zig:276:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha1, "da39a3ee5e6b4b0d3255bfef95601890afd80709", "");
             ^
./lib/std/crypto/sha1.zig:286:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("da39a3ee5e6b4b0d3255bfef95601890afd80709", out[0..]);
             ^
./lib/std/crypto/sha2.zig:296:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha224, "d14a028c2a3a2bc9476102bb288234c415a2b01f828ea62ac5b3e42f", "");
             ^
./lib/std/crypto/sha2.zig:306:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("d14a028c2a3a2bc9476102bb288234c415a2b01f828ea62ac5b3e42f", out[0..]);
             ^
./lib/std/crypto/sha2.zig:322:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha256, "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", "");
             ^
./lib/std/crypto/sha2.zig:332:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855", out[0..]);
             ^
./lib/std/crypto/sha2.zig:687:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha384, h1, "");
             ^
./lib/std/crypto/sha2.zig:702:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/sha2.zig:721:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha512, h1, "");
             ^
./lib/std/crypto/sha2.zig:736:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/sha3.zig:180:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha3_224, "6b4e03423667dbb73b6e15454f0eb1abd4597f9a1b078e3f5b5a6bc7", "");
             ^
./lib/std/crypto/sha3.zig:190:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("6b4e03423667dbb73b6e15454f0eb1abd4597f9a1b078e3f5b5a6bc7", out[0..]);
             ^
./lib/std/crypto/sha3.zig:206:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha3_256, "a7ffc6f8bf1ed76651c14756a061d662f580ff4de43b49fa82d80a4b80f8434a", "");
             ^
./lib/std/crypto/sha3.zig:216:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("a7ffc6f8bf1ed76651c14756a061d662f580ff4de43b49fa82d80a4b80f8434a", out[0..]);
             ^
./lib/std/crypto/sha3.zig:242:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha3_384, h1, "");
             ^
./lib/std/crypto/sha3.zig:255:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/sha3.zig:273:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Sha3_512, h1, "");
             ^
./lib/std/crypto/sha3.zig:286:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(h1, out[0..]);
             ^
./lib/std/crypto/sha3.zig:312:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Keccak_256, "c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470", "");
             ^
./lib/std/crypto/sha3.zig:318:14: error: container 'crypto.test' has no member called 'assertEqualHash'
    try htest.assertEqualHash(Keccak_512, "0eab42de4c3ceb9235fc91acffe746b29c29a8c366b7c60e4e67c466f36a4304c00fa9caf9d87976ba469bcbe06713b435f091ef2769fb160cdab33d3670680e", "");
             ^
./lib/std/crypto/gimli.zig:286:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("1C9A03DC6A5DDC5444CFC6F4B154CFF5CF081633B2CEA4D7D0AE7CCFED5AAA44", &md);
             ^
./lib/std/crypto/gimli.zig:294:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("404C130AF1B9023A7908200919F690FFBB756D5176E056FFDE320016A37C7282", &md);
             ^
./lib/std/crypto/gimli.zig:302:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("A8F4FA28708BDA7EFB4C1914CA4AFA9E475B82D588D36504F87DBB0ED9AB3C4B", &md);
             ^
./lib/std/crypto/gimli.zig:459:18: error: container 'crypto.test' has no member called 'assertEqual'
        try htest.assertEqual("", &ct);
                 ^
./lib/std/crypto/hmac.zig:82:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("74e6f7298a9c2d168935f58c001bad88", out[0..]);
             ^
./lib/std/crypto/hmac.zig:91:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("fbdb1d1b18aa6c08324b7d64b71fb76370690e1d", out[0..]);
             ^
./lib/std/crypto/hmac.zig:100:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("b613679a0814d9ec772f95d778c35fc5ff1697c493715653c6c712144292c5ad", out[0..]);
             ^
./lib/std/crypto/25519/x25519.zig:176:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("90e7595fc89e52fdfddce9c6a43d74dbf6047025ee0462d2d172e8b6a2841d6e", &mont_kp.secret_key);
             ^
./lib/std/crypto/25519/edwards25519.zig:538:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("0691eee3cf70a0056df6bfa03120635636581b5c4ea571dfc680f78c7e0b4137", p.toBytes()[0..]);
             ^
./lib/std/crypto/25519/edwards25519.zig:548:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("31558a26887f23fb8218f143e69d5f0af2e7831130bd5b432ef23883b895839a", p.toBytes()[0..]);
             ^
./lib/std/crypto/hkdf.zig:62:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("077709362c2e32df0ddc3f0dc47bba6390b6c73bb50f9c3122ec844ad7c2b3e5", &prk);
             ^
./lib/std/crypto/ghash.zig:325:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("889295fa746e8b174bf4ec80a65dea41", &out);
             ^
./lib/std/crypto/pbkdf2.zig:165:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(expected, dk[0..]);
             ^
./lib/std/crypto/pbkdf2.zig:180:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(expected, dk[0..]);
             ^
./lib/std/crypto/pbkdf2.zig:195:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(expected, dk[0..]);
             ^
./lib/std/crypto/pbkdf2.zig:230:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(expected, dk[0..]);
             ^
./lib/std/crypto/pbkdf2.zig:245:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual(expected, dk[0..]);
             ^
./lib/std/crypto/salsa20.zig:559:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("30ff9933aa6534ff5207142593cd1fca4b23bdd8", c[0..]);
             ^
./lib/std/crypto/salsa20.zig:626:14: error: container 'crypto.test' has no member called 'assertEqual'
    try htest.assertEqual("b05760e217288ba079caa2fd57fd3701784974ffcfda20fe523b89211ad8af065a6eb37cdb29d51aca5bd75dafdd21d18b044c54bb7c526cf576c94ee8900f911ceab0147e82b667a28c52d58ceb29554ff45471224d37b03256b01c119b89ff6d36855de8138d103386dbc9d971f52261", &ciphertext);
             ^

```

**Comment by hnakamur** [comment]
`@embedFile` errors are gone after I copied `lib/std/compress/rfc*` files from the zig git repository.

```
lxc file push lib/std/compress/rfc* bionic-vm/root/zig-linux-x86_64-0.9.0-dev.1815+20e19e75f/lib/std/compress/
```

The errors like `error: container 'crypto.test' has no member called 'assertEqual'` are still there.

**Comment by hnakamur** [comment]
The tests passed using the std lib source from the git repository.

```
cd
git clone https://github.com/ziglang/zig
cd zig
../zig-linux-x86_64-0.9.0-dev.1815+20e19e75f/zig test lib/std/std.zig --zig-lib-dir lib --main-pkg-path lib/std
```

The output:
```
Test [284/2077] os.linux.io_uring.test "write/read"... SKIP
Test [286/2077] os.linux.io_uring.test "openat"... SKIP
Test [287/2077] os.linux.io_uring.test "close"... SKIP
Test [288/2077] os.linux.io_uring.test "accept/connect/send/recv"... SKIP
Test [291/2077] os.linux.io_uring.test "timeout_remove"... SKIP
Test [292/2077] os.linux.io_uring.test "timeout_link_chain1"... SKIP
Test [293/2077] os.linux.io_uring.test "fallocate"... SKIP
Test [294/2077] os.linux.io_uring.test "statx"... SKIP
Test [295/2077] os.linux.io_uring.test "accept/connect/recv/cancel"... SKIP
Test [296/2077] os.linux.io_uring.test "register_files_update"... SKIP
Test [372/2077] packed_int_array.test "PackedIntArray"... SKIP
Test [376/2077] packed_int_array.test "PackedIntSlice"... SKIP
Test [377/2077] packed_int_array.test "PackedIntSlice of PackedInt(Array/Slice)"... SKIP
Test [431/2077] Progress.test "basic functionality"... SKIP
Test [763/2077] event.loop.test "std.event.Loop - basic"... SKIP
Test [764/2077] event.loop.test "std.event.Loop - runDetached"... SKIP
Test [765/2077] event.loop.test "std.event.Loop - sleep"... SKIP
Test [821/2077] io.c_writer.test ""... SKIP
Test [873/2077] fs.watch.test "write a file, watch it, write it again, delete it"... SKIP
Test [878/2077] os.test.test "readlink on Windows"... SKIP
Test [1040/2077] event.channel.test "std.event.Channel"... SKIP
Test [1041/2077] event.channel.test "std.event.Channel wraparound"... SKIP
Test [1042/2077] event.future.test "std.event.Future"... SKIP
Test [1043/2077] event.group.test "std.event.Group"... SKIP
Test [1045/2077] event.lock.test "std.event.Lock"... SKIP
Test [1046/2077] event.rwlock.test "std.event.RwLock"... SKIP
Test [1047/2077] event.wait_group.test "basic WaitGroup usage"... SKIP
Test [1407/2077] net.test.test "listen on a port, send bytes, receive bytes"... SKIP
Test [1408/2077] net.test.test "listen on ipv4 try connect on ipv6 then ipv4"... SKIP
Test [1474/2077] crypto.25519.x25519.test "x25519 rfc7748 1,000 iterations"... SKIP
Test [1475/2077] crypto.25519.x25519.test "x25519 rfc7748 1,000,000 iterations"... SKIP
Test [1503/2077] crypto.scrypt.test "kdf"... SKIP
Test [1504/2077] crypto.scrypt.test "kdf rfc 1"... SKIP
Test [1505/2077] crypto.scrypt.test "kdf rfc 2"... SKIP
Test [1506/2077] crypto.scrypt.test "kdf rfc 3"... SKIP
Test [1507/2077] crypto.scrypt.test "kdf rfc 4"... SKIP
Test [1508/2077] crypto.scrypt.test "password hashing (crypt format)"... SKIP
Test [1509/2077] crypto.scrypt.test "strHash and strVerify"... SKIP
Test [1510/2077] crypto.scrypt.test "unix-scrypt"... SKIP
Test [1516/2077] crypto.pbkdf2.test "RFC 6070 16,777,216 iterations"... SKIP
Test [1519/2077] crypto.pbkdf2.test "Very large dk_len"... SKIP
Test [2064/2077] zig.parser_test.test "recovery: missing block after suspend"... SKIP
2035 passed; 42 skipped; 0 failed.
```

```
root@bionic-vm:~/zig# echo $?
0
```

The commit I used is:

```
root@bionic-vm:~/zig# git log -1
commit 2873e19366400b2f11d3a99b10ea17e4f06c6b4b (HEAD -> master, origin/master, origin/HEAD)
Author: Jakub Konka <kubkon@jakubkonka.com>
Date:   Mon Nov 29 21:05:19 2021 +0100

    macho: ignore undefined symbols marked for discarding

    If a symbol is undefined after we tried resolving it in static and
    dynamic libraries, and it is annotated with N_DESC_DISCARDED flag,
    we simply ignore it rather than flagging an undefined symbol error.
```

**Comment by andrewrk** [comment]
Spotted on the CI: https://ci.ziglang.org/ziglang/zig/515/1/4

```
Test [245/2077] os.linux.io_uring.test "std-native-ReleaseSafe-c-multi timeout_link_chai... thread 57050 panic: reached unreachable code
error: the following test command crashed:
/workspace/zig-cache/o/84f54c420fa45497d421d687f653af91/test /workspace/_debug/staging/bin/zig
The following command exited with error code 1:
/workspace/_debug/staging/bin/zig test /workspace/lib/std/std.zig -lc --test-name-prefix std-native-ReleaseSafe-c-multi  -OReleaseSafe --cache-dir /workspace/zig-cache --global-cache-dir /root/.cache/zig --name test -fno-single-threaded -I /workspace/test --zig-lib-dir /workspace/lib 
```

**Comment by hnakamur** [comment]
Thanks again for finding the problem and fixing it with a workaround.
Is there anything I can do now?

**Comment by andrewrk** [comment]
My goal is to make sure we have no flaky test failures. So I will put a workaround into this test case and I wanted to give you a chance to remove the workaround in some better way. If there is no better way than to skip the test when this problem occurs, then we can close this issue.

**Comment by hnakamur** [comment]
The old test "timeout_link_chain1" was ported from liburing test_timeout_link_chain1
https://github.com/axboe/liburing/blob/509873c4454012c5810c728695c21911c82acdc4/test/link-timeout.c#L539-L628
However it turns out that both fails with EBADF (-9) on Linux kernel 5.4.

I created a pull request to [Fix test for io_uring link_timeout (fixes #10247) by hnakamur · Pull Request #10257 · ziglang/zig](https://github.com/ziglang/zig/pull/10257).

The this new test skips properly on Linux kernel 5.4
and passes on Linux kernel 5.11.

**Comment by andrewrk** [comment]
Thank you!

**Comment by andrewrk** [comment]
Fixed in 0714832c21169493a0a857237fbface612e3d5e0

**Comment by hnakamur** [comment]
Thank you too!
